### DISCUSSION TITLE
Inserting and Removing DMX Cable - Causing "Unrecoverable" DMX_ERR_IMPROPER_SLOT

### DISCUSSION THREAD

**Discussion by Ben-Rosenfeld-VPS** [someweisguy/esp_dmx#174] [General]
Hi,

We see the following when inserting a DMX cable from a lighting console that is actively sending DMX:

![image](https://github.com/user-attachments/assets/1ab4638f-0091-4e06-9726-7cc56d698ff7)

In the above image the data in line 6 should have matched the data in line 1.

In software we ultimately perform the following logic:

A. Call to dmx_receive with a wait ticks of 10ms
B. If there is no data for 100ms then the DMX State is changed to 0 (IDLE); this was seen in line 2 in the image
C. If there is data then switch on packet.err
D. When packet.err is set to DMX_ERR_IMPROPER_SLOT log; this was seen in line 3 and 4 in the image
E. Otherwise when packet.err is set to DMX_OK we log the first 10 bytes of the data; as shown in the image line 1 and 6 log data

What is interesting is that after the DMX_ERR_IMPROPER_SLOT, the dmx_receive will always return all 0's for the data. If we then insert and remove the cable multiple times, it will sometimes eventually recover and send the actual data.

**Comment by putzlicht** [comment]
Hi Ben,

just in case that you are still struggling with this - use dmx_receive_num() instead of dmx_receive().
dmx_receive() has a nasty function. It remembers the number of slots that it has received (until a break) for the next frame it is waiting for. Unfortunately it can't decide if a drop of slot numbers is that what you want or if the drop happened accidentally.
With removing and inserting the dmx cable I was able to drop the slot numbers in some seconds from 512 to 1.
 As it depends on the actual position in the transmitted frame it varies from try to try.
(the function could check for more than one frame if the number of the received slots is constant before it starts to adapt)
dmx_receive_num(port, &packet, 128, timeout) for example always waits for 128 Slots before it returns. For me this works flawless. Just drop the packets with errors and check if the startcode in your data[0] == SC_DMX so that you can discard faulty packets you received during a dmx cable handling.
And don't decrease the timeout to much. Otherwise the read function will return to early and can't get all the Slots you are waiting for. Maximum dmx frame rate is 44. Therefore a frame of 512 Slots needs around 23 milliseconds to be transmitted.
